### PR TITLE
Fix address computation in Slice constructors and add tests

### PR DESCRIPTION
**Pull request by raghavsethi**

### CODE REVIEW

File: src/test/java/io/airlift/slice/TestSlices.java [hunk 1]
```diff
@@ -67,6 +79,58 @@ private static void testWrapping(ByteBuffer buffer)
...
     }
 
+    @Test
+    public void testWrapAllTypes()
+    {
+        boolean[] booleanArray = {true, false, false, true, false};
+        assertEquals(wrappedBooleanArray(booleanArray).getByte(0) == 1, booleanArray[0]);
+        SliceInput input = wrappedBooleanArray(booleanArray).getInput();
+        input.readBoolean();
+        assertEquals(input.readBoolean(), booleanArray[1]);
```
Comment (haozhun) [inline src/test/java/io/airlift/slice/TestSlices.java]: These `SliceInput` tests doesn't seem relevant to the changes your made. In addition, these cases are covered in `AbstractSliceInputTest` already.

File: src/main/java/io/airlift/slice/Slice.java [hunk 2]
```diff
@@ -145,89 +148,107 @@ public static Slice toUnsafeSlice(ByteBuffer byteBuffer)
...
+     * @param length the number of array positions to include in the slice
      */
     Slice(boolean[] base, int offset, int length)
     {
         requireNonNull(base, "base is null");
         checkPositionIndexes(offset, offset + length, base.length);
 
         this.base = base;
-        this.address = ARRAY_BOOLEAN_BASE_OFFSET + offset;
+        this.address = ARRAY_BOOLEAN_BASE_OFFSET + (offset * ARRAY_BOOLEAN_INDEX_SCALE);
```
Comment (haozhun) [inline src/main/java/io/airlift/slice/Slice.java]: nit: remove brackets to be consistent with 2 lines below.

File: src/test/java/io/airlift/slice/TestSlices.java (hunk 1 above)
Comment (raghavsethi) [inline src/test/java/io/airlift/slice/TestSlices.java]: That's a fair point.

File: src/test/java/io/airlift/slice/TestSlices.java [hunk 3]
```diff
@@ -67,6 +79,40 @@ private static void testWrapping(ByteBuffer buffer)
         }
     }
 
+    @Test
+    public void testWrapAllTypes()
```
Comment (electrum) [inline src/test/java/io/airlift/slice/TestSlices.java]: Add byte array as well

File: src/test/java/io/airlift/slice/TestSlices.java [hunk 4]
```diff
@@ -67,6 +79,40 @@ private static void testWrapping(ByteBuffer buffer)
...
+        assertEquals(wrappedShortArray(shortArray).getShort(0), shortArray[0]);
+        assertEquals(wrappedShortArray(shortArray, 1, 4).getShort(0), shortArray[1]);
+        assertEquals(wrappedShortArray(shortArray).getShort((4) * SIZE_OF_SHORT), shortArray[4]);
+
+        int[] intArray = {0, 1, 2, 3, 4};
+        assertEquals(wrappedIntArray(intArray).getInt(0), intArray[0]);
+        assertEquals(wrappedIntArray(intArray, 1, 4).getInt(0), intArray[1]);
+        assertEquals(wrappedIntArray(intArray).getInt((4) * SIZE_OF_INT), intArray[4]);
+
+        float[] floatArray = {0f, 1f, 2f, 3f, 4f};
```
Comment (electrum) [inline src/test/java/io/airlift/slice/TestSlices.java]: Move this after long so float/double are togther

File: src/main/java/io/airlift/slice/Slice.java [hunk 5]
```diff
@@ -130,6 +130,9 @@ public static Slice toUnsafeSlice(ByteBuffer byteBuffer)
 
     /**
      * Creates a slice over the specified array range.
+     *
+     * @param offset the array position at which the slice begins
```
Comment (electrum) [inline src/main/java/io/airlift/slice/Slice.java]: Put the Javadoc on the public factory methods

File: src/test/java/io/airlift/slice/TestSlices.java [hunk 6]
```diff
@@ -67,6 +79,40 @@ private static void testWrapping(ByteBuffer buffer)
         }
     }
 
+    @Test
+    public void testWrapAllTypes()
+    {
+        boolean[] booleanArray = {true, false, false, true, false};
```
Comment (electrum) [inline src/test/java/io/airlift/slice/TestSlices.java]: Make all these have 6 entries, so the slice is in the middle

File: src/test/java/io/airlift/slice/TestSlices.java [hunk 7]
```diff
@@ -67,6 +79,40 @@ private static void testWrapping(ByteBuffer buffer)
...
+        assertEquals(wrappedBooleanArray(booleanArray).getByte(0) == 1, booleanArray[0]);
+        assertEquals(wrappedBooleanArray(booleanArray, 1, 4).getByte(0) == 1, booleanArray[1]);
+        assertEquals(wrappedBooleanArray(booleanArray).getByte((4) * SIZE_OF_BYTE) == 1, booleanArray[4]);
```
Comment (electrum) [inline src/test/java/io/airlift/slice/TestSlices.java]: No need for parens

File: src/test/java/io/airlift/slice/TestSlices.java (hunk 6 above)
Comment (electrum) [inline src/test/java/io/airlift/slice/TestSlices.java]: Add a test that verifies the `length()` of the slice

File: src/test/java/io/airlift/slice/TestSlices.java [hunk 8]
```diff
@@ -67,6 +80,52 @@ private static void testWrapping(ByteBuffer buffer)
         }
     }
 
+    @Test
+    public void testWrapAllTypes()
+    {
+        boolean[] booleanArray = {true, false, false, true, false, true};
+        assertEquals(wrappedBooleanArray(booleanArray).getByte(0) == 1, booleanArray[0]);
+        assertEquals(wrappedBooleanArray(booleanArray, 2, 4).getByte(0) == 1, booleanArray[1]);
```
Comment (electrum) [inline src/test/java/io/airlift/slice/TestSlices.java]: Can you make all these `1, 4` so the slice is in the middle of the array?

Also, with the current code, this should be `booleanArray[2]`, but it will be correct after updating the slice.
